File: kafkaStreamsModelServerKVStore/src/main/java/com/lightbend/standard/modelserver/DataProcessor.java
package com.lightbend.standard.modelserver;

import com.lightbend.configuration.kafka.ApplicationKafkaParameters;
import com.lightbend.model.Winerecord;
import com.lightbend.model.ModelServingInfo;
import com.lightbend.standard.modelserver.store.StoreState;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;
import org.apache.kafka.streams.state.KeyValueStore;

import java.util.Objects;
import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 * used
 * https://github.com/bbejeck/kafka-streams/blob/master/src/main/java/bbejeck/processor/stocks/StockSummaryProcessor.java
 */
public class DataProcessor extends AbstractProcessor<byte[], Optional<Winerecord.WineRecord>> {

    private KeyValueStore<Integer, StoreState> modelStore;

    @Override
    public void process(byte[] key, Optional<Winerecord.WineRecord> dataRecord) {

        StoreState state = modelStore.get(ApplicationKafkaParameters.STORE_ID);
        if(state == null)
            state = new StoreState();
        if(state.getNewModel() != null){
            // update the model
            if(state.getCurrentModel() != null)
                state.getCurrentModel().cleanup();
            state.setCurrentModel(state.getNewModel());
            state.setCurrentServingInfo(new ModelServingInfo( state.getNewServingInfo().getName(),
                     state.getNewServingInfo().getDescription(), System.currentTimeMillis()));
            state.setNewServingInfo(null);
            state.setNewModel(null);
        }
        // Actually score
        if( state.getCurrentModel() == null) {
            // No model currently
            System.out.println("No model available - skipping");
        }
        else{
            // Score the model
            long start = System.currentTimeMillis();
            double quality = (double)  state.getCurrentModel().score(dataRecord.get());
            long duration = System.currentTimeMillis() - start;
             state.getCurrentServingInfo().update(duration);
            System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
        }
        modelStore.put(ApplicationKafkaParameters.STORE_ID, state);
    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = (KeyValueStore<Integer, StoreState>) context.getStateStore(ApplicationKafkaParameters.STORE_NAME);
        Objects.requireNonNull(modelStore, "State store can't be null");

    }
}

File: kafkaStreamsModelServerKVStore/src/main/scala/com/lightbend/scala/standard/modelserver/DataProcessor.scala
package com.lightbend.scala.standard.modelserver.scala

import java.util.Objects

import com.lightbend.model.winerecord.WineRecord
import com.lightbend.java.configuration.kafka.ApplicationKafkaParameters
import com.lightbend.scala.standard.modelserver.scala.store.StoreState
import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext, ProcessorSupplier}
import org.apache.kafka.streams.state.KeyValueStore

import scala.util.Try

class DataProcessor extends AbstractProcessor[Array[Byte], Try[WineRecord]] with ProcessorSupplier[Array[Byte], Try[WineRecord]]{

  private var modelStore = null.asInstanceOf[KeyValueStore[Integer, StoreState]]

  import ApplicationKafkaParameters._

  override def process(key: Array[Byte], dataRecord: Try[WineRecord]): Unit = {

    var state = modelStore.get(STORE_ID)
    if (state == null) state = new StoreState

    state.newModel match {
      case Some(model) => {
        // close current model first
        state.currentModel match {
          case Some(m) => m.cleanup()
          case _ =>
        }
        // Update model
        state.currentModel = Some(model)
        state.currentState = state.newState
        state.newModel = None
      }
      case _ =>
    }
    state.currentModel match {
      case Some(model) => {
        val start = System.currentTimeMillis()
        val quality = model.score(dataRecord.get.asInstanceOf[AnyVal]).asInstanceOf[Double]
        val duration = System.currentTimeMillis() - start
        println(s"Calculated quality - $quality calculated in $duration ms")
        state.currentState.get.incrementUsage(duration)
      }
      case _ => {
        println("No model available - skipping")
      }
    }
    modelStore.put(STORE_ID, state)
  }

  override def init(context: ProcessorContext): Unit = {
    modelStore = context.getStateStore(STORE_NAME).asInstanceOf[KeyValueStore[Integer, StoreState]]
    Objects.requireNonNull(modelStore, "State store can't be null")
  }

  override def get() = new DataProcessor()
}
